src/data/value/into.rs

```diff
@@ -64,6 +66,7 @@ impl TryInto<bool> for &Value {
                 "FALSE" => false,
                 _ => return Err(ValueError::ImpossibleCast.into()),
             },
+            Value::Decimal(value) => !value.is_zero(),
```

Comment:
Consider using a constant variables: 
1. <https://docs.rs/rust_decimal/1.17.0/rust_decimal/struct.Decimal.html#associatedconstant.ZERO> 
2. <https://docs.rs/rust_decimal/1.17.0/rust_decimal/struct.Decimal.html#associatedconstant.ONE>

Reply:
I tested casting to boolean with PostgreSQL v.14  

Firstly, I tried to cast Decimal to boolean. 
```sql
create table hello (a decimal);
insert into hello values (1);
insert into hello values (0);
insert into hello values (3);
select CAST(a as boolean) from hello;
ERROR:  cannot cast type numeric to boolean
LINE 1: select CAST(a as boolean) from hello;
```
It seems that PostgreSQL does not support cast Decimal into boolean. 

Then, I tried cast INT into boolean. 
```sql 
create table int_test (a int);
insert into int_test values(1);
insert into int_test values(2);
insert into int_test values(3);
insert into int_test values(0);
insert into int_test values(-1);
insert into int_test values(-2);
select a, CAST(a as boolean) casted from int_test;
 a  | casted 
----+--------
  1 | t
  2 | t
  3 | t
  0 | f
 -1 | t
 -2 | t
(6 rows)
```

It seems that postgres casts Int values which are not zero  to `true` and only `0`  to  `false`. 
I "feel" this is quite logical since if there is any value, it represents `true` and only `0` represents `false`. 

I've searched why postgresql does not support cast `Decimal` (or `Numeric`) to `boolean`, but cannot find any document about it. 
For Mysql, does not support CAST to Boolean from any data type. 

By the way, I found some other SQL or Databases supporting cast  `Decimal` to `boolean`
https://docs.mongodb.com/manual/reference/operator/aggregation/convert/#std-label-convert-to-bool
https://impala.apache.org/docs/build/html/topics/impala_boolean.html
https://docs.databricks.com/sql/language-manual/functions/cast.html

If we choose to follow PostgreSQL for reference, I think we should change our `try_into` function for I64. 
If its value is not zero, it returns `true`, else it returns `false`. 
and for Decimal and F64, we can just make it `impossible cast`.  (I've also tested cast `Float` to boolean and it doesn't support) 
and for String, we can add some other string values that postgre supports cast 
"true"s | "false"s
---- | ----
true | false
yes | no
on | off
1 | 0

Then codes will be like below 

``` rust
...
Value::I64(value) => !value.is_zero(),
Value::Str(value) => match value.to_uppercase().as_str() {
                "TRUE" | "YES" | "ON" | "1" => true,
                "FALSE"  | "NO" | "OFF" | "0" => false,
                _ => return Err(ValueError::ImpossibleCast.into()),
            },
Value::Decimal(_) 
| Value::F64(_) 
| Value::Date(_)
| Value::Timestamp(_)
| Value::Time(_)
| Value::Interval(_)
| Value::Uuid(_)
| Value::Map(_)
| Value::List(_)
| Value::Null => return Err(ValueError::ImpossibleCast.into()),
... 
```

Or, we can choose to follow other databases and make it possible to cast any numeric types(int, float, decimal) to boolean.

Reply:
Thanks for the sharing, the examples are really helpful to decide what rule to choose.
My idea is this.
We don't need to follow the rule from other legacy dbs, and I'm for making our own rule at this point.
There's a single point that I want to follow for our casting rule.
* Type casting between boolean and (numeric or string) values would be better to have 1:1 relation.

It would be more strict but consistent and straightforward to understand, I think.
If we accept multiple string values for boolean (`TRUE` | `YES` | `ON` | `1`), that multiple values converge into a single `Bool(true)` value and this action is not reversible because `Bool(true)` to `Str` can return only a single value (`TRUE`).
Then, we need to explain two ways differently, `Str` to `Bool` and `Bool` to `Str`. Users also need to remember both of them.
1. `Str` to `Bool` - `"TRUE" | "YES" | "ON" | "1"`
2. `Bool` to `Str` - `true -> "TRUE"`

However, if we only accept a single value `TRUE` for `Bool(true)`, it becomes more simple. Now, there's only one thing we need to explain and what users need to know.
1. `"TRUE" <---> true`

It can be also applied to cast between numeric and boolean.

* non-zero to true, then
1. `Numeric` to `Bool` - `1, 2, 3, ... -> true`
2. `Bool` to `Numeric` - `true -> 1`
* only accepting single `1` value for true
1. `1 <---> true`

Then...

| `Boolean(true)` | `Boolean(false)` |
| --- | --- |
| `Numeric(1)` | `Numeric(0)` |
| `String("TRUE")` | `String("FALSE")` |

How do you think of this?

Reply:
Currently, in the conversion between string and boolean, we handle string value in case-insensitively so it is actually not that really 1:1 but... I was just not sure about it so.. current str to bool accepts multiple string values `"True" | "true" | 
"tRue" | "TRUE" | ...`.

Reply:
As long as the rationale for our decision is clear, it is better to go our own way. 

Let's just stick to this table. 

Boolean(true) | Boolean(false)
-- | --
Numeric(1) | Numeric(0)
String("TRUE") | String("FALSE")

I also thought about the actual use case, but it seems clear from the user's point of view that only `1` is converted to `true` rather than **all numbers except 0** are converted to `true`.

I will reflect this in this PR. (using constant variables which @ever0de commented) 

Thanks a lot!